README.rst
==============
Ruuvi Decoders
==============


.. image:: https://img.shields.io/pypi/v/ruuvi_decoders.svg
        :target: https://pypi.python.org/pypi/ruuvi_decoders

.. image:: https://img.shields.io/travis/ruuvi-friends/ruuvi-decoders.svg
        :target: https://travis-ci.com/ruuvi-friends/ruuvi-decoders

.. image:: https://pyup.io/repos/github/sergioisidoro/ruuvi_decoders/shield.svg
     :target: https://pyup.io/repos/github/sergioisidoro/ruuvi_decoders/
     :alt: Updates


Decoders for raw ruuvi tag data payloads


* Free software: MIT license
* Documentation: https://ruuvi-decoders.readthedocs.io.


Features
--------

* Decode raw data from RuuviTag Bluetooh advertisements, or from Ruuvi gateway requests
* Support for Ruuvitag Data Format 2, 3 and 5
* Support for ULRDecoder (Eddystone)

Credits
-------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

0.2.0 (2021-07-29)
------------------

* Internal refactoring

0.1.0 (2021-04-25)
------------------

* First release on PyPI.


tests/__init__.py
"""Unit test package for ruuvi_decoders."""


tests/test_ruuvi_decoders.py
#!/usr/bin/env python

from unittest import TestCase

from ruuvi_decoders import get_decoder, UrlDecoder, Df3Decoder, Df5Decoder

# pylint: disable=line-too-long

class TestDecoder(TestCase):

    def test_getcorrectdecoder(self):
        dec = get_decoder(2)
        self.assertIsInstance(dec, UrlDecoder)
        dec = get_decoder(3)
        self.assertIsInstance(dec, Df3Decoder)

    def test_decode_is_valid(self):
        decoder = UrlDecoder()
        data = decoder.decode_data('AjwYAMFc')

        self.assertEqual(data['temperature'], 24)
        self.assertEqual(data['pressure'], 995)
        self.assertEqual(data['humidity'], 30)
        self.assertEqual(data['identifier'], None)

    def test_decode_is_valid_case2(self):
        decoder = UrlDecoder()
        data = decoder.decode_data('AjgbAMFc')

        self.assertEqual(data['temperature'], 27)
        self.assertEqual(data['pressure'], 995)
        self.assertEqual(data['humidity'], 28)
        self.assertEqual(data['identifier'], None)

    def test_decode_is_valid_weatherstation_2017_04_12(self):
        decoder = UrlDecoder()
        data = decoder.decode_data('AjUX1MAw0')

        self.assertEqual(data['temperature'], 25.12)
        self.assertEqual(data['pressure'], 992.0)
        self.assertEqual(data['humidity'], 26.5)
        self.assertEqual(data['identifier'], '0')

    def test_df3decode_is_valid(self):
        decoder = Df3Decoder()
        data = decoder.decode_data('03291A1ECE1EFC18F94202CA0B5300000000BB')

        self.assertEqual(data['temperature'], 26.3)
        self.assertEqual(data['pressure'], 1027.66)
        self.assertEqual(data['humidity'], 20.5)
        self.assertEqual(data['battery'], 2899)
        self.assertNotEqual(data['acceleration'], 0)
        self.assertEqual(data['acceleration_x'], -1000)
        self.assertNotEqual(data['acceleration_y'], 0)
        self.assertNotEqual(data['acceleration_z'], 0)

        data = decoder.decode_data('03291A1ECE1EFC18F94202CA0B53BB')
        self.assertEqual(data['temperature'], 26.3)
        self.assertEqual(data['pressure'], 1027.66)
        self.assertEqual(data['humidity'], 20.5)
        self.assertEqual(data['battery'], 2899)
        self.assertNotEqual(data['acceleration'], 0)
        self.assertEqual(data['acceleration_x'], -1000)
        self.assertNotEqual(data['acceleration_y'], 0)
        self.assertNotEqual(data['acceleration_z'], 0)

    def test_df3decode_is_valid_notNone(self):
        test_cases = [
            '1502010611FF990403411540C84AFC72FE2FFFC50B89C6',
            '1502010611FF990403411544C850FC72FE2FFFC60B89B9',
            '1502010611FF990403411540C855FC72FE2FFFC30B83C7',
            '1502010611FF990403411539C842FC72FE2FFFC60B89C5',
            '1502010611FF990403421534C813FC72FE2FFFC50B8FD5',
            '1502010611FF990403441536C810FC72FE2FFFC70B83C7',
        ]
        decoder = Df3Decoder()
        for x in test_cases:
            data = decoder.decode_data(x)
            self.assertIsNotNone(data)

    def test_df3decode_is_valid_max_values(self):
        decoder = Df3Decoder()
        humidity = 'C8'
        temp = '7F63'
        pressure = 'FFFF'
        accX = '03E8'
        accY = '03E8'
        accZ = '03E8'
        batt = 'FFFF'
        data = decoder.decode_data(
            '03{humidity}{temp}{pressure}{accX}{accY}{accZ}{batt}00000000BB'.format(
                humidity=humidity,
                temp=temp,
                pressure=pressure,
                accX=accX,
                accY=accY,
                accZ=accZ,
                batt=batt))

        self.assertEqual(data['temperature'], 127.99)
        self.assertEqual(data['pressure'], 1155.35)
        self.assertEqual(data['humidity'], 100.0)
        self.assertEqual(data['battery'], 65535)
        self.assertEqual(data['acceleration_x'], 1000)
        self.assertEqual(data['acceleration_y'], 1000)
        self.assertEqual(data['acceleration_z'], 1000)
        self.assertNotEqual(data['acceleration'], 0)

    def test_df3decode_is_valid_min_values(self):
        decoder = Df3Decoder()
        humidity = '00'
        temp = 'FF63'
        pressure = '0000'
        accX = 'FC18'
        accY = 'FC18'
        accZ = 'FC18'
        batt = '0000'
        data = decoder.decode_data(
            '03{humidity}{temp}{pressure}{accX}{accY}{accZ}{batt}00000000BB'.format(
                humidity=humidity,
                temp=temp,
                pressure=pressure,
                accX=accX,
                accY=accY,
                accZ=accZ,
                batt=batt))

        self.assertEqual(data['temperature'], -127.99)
        self.assertEqual(data['pressure'], 500.0)
        self.assertEqual(data['humidity'], 0.0)
        self.assertEqual(data['battery'], 0)
        self.assertEqual(data['acceleration_x'], -1000)
        self.assertEqual(data['acceleration_y'], -1000)
        self.assertEqual(data['acceleration_z'], -1000)
        self.assertNotEqual(data['acceleration'], 0)

    def test_df5decode_is_valid(self):
        decoder = Df5Decoder()
        data_format = '05'
        temp = '12FC'
        humidity = '5394'
        pressure = 'C37C'
        accX = '0004'
        accY = 'FFFC'
        accZ = '040C'
        power_info = 'AC36'
        movement_counter = '42'
        measurement_sequence = '00CD'
        mac = 'CBB8334C884F'
        data = decoder.decode_data(
            '{format}{temp}{humidity}{pressure}{accX}{accY}{accZ}{power_info}{movement_counter}{measurement_sequence}{mac}'.format(
                format=data_format,
                humidity=humidity,
                temp=temp,
                pressure=pressure,
                accX=accX,
                accY=accY,
                accZ=accZ,
                power_info=power_info,
                movement_counter=movement_counter,
                measurement_sequence=measurement_sequence,
                mac=mac))

        self.assertEqual(data['temperature'], 24.30)
        self.assertEqual(data['humidity'], 53.49)
        self.assertEqual(data['pressure'], 1000.44)
        self.assertEqual(data['acceleration_x'], 4)
        self.assertEqual(data['acceleration_y'], -4)
        self.assertEqual(data['acceleration_z'], 1036)
        self.assertEqual(data['tx_power'], 4)
        self.assertEqual(data['battery'], 2977)
        self.assertEqual(data['movement_counter'], 66)
        self.assertEqual(data['measurement_sequence_number'], 205)
        self.assertEqual(data['mac'], 'cbb8334c884f')


ruuvi_decoders/__init__.py
"""Top-level package for Ruuvi Decoders."""

__author__ = """Sergio Isidoro"""
__email__ = 'smaisidoro@gmail.com'
__version__ = '0.1.0'
import logging
from ruuvi_decoders.url_decoder import UrlDecoder
from ruuvi_decoders.df3_decoder import Df3Decoder
from ruuvi_decoders.df5_decoder import Df5Decoder

log = logging.getLogger(__name__)


def get_decoder(data_type):
    """
    Get correct decoder for Data Type.

    Returns:
        object: Data decoder
    """
    if data_type == 2:
        log.warning("DATA TYPE 2 IS OBSOLETE. UPDATE YOUR TAG")
        # https://github.com/ruuvi/ruuvi-sensor-protocols/blob/master/dataformat_04.md
        return UrlDecoder()
    if data_type == 4:
        log.warning("DATA TYPE 4 IS OBSOLETE. UPDATE YOUR TAG AND SWITCH MODE")
        # https://github.com/ruuvi/ruuvi-sensor-protocols/blob/master/dataformat_04.md
        return UrlDecoder()
    if data_type == 3:
        log.warning("DATA TYPE 3 IS DEPRECATED - UPDATE YOUR TAG")
        # https://github.com/ruuvi/ruuvi-sensor-protocols/blob/master/dataformat_03.md
        return Df3Decoder()
    return Df5Decoder()


ruuvi_decoders/df3_decoder.py
from __future__ import division

import math
import logging
import struct

log = logging.getLogger(__name__)


class Df3Decoder(object):
    """
    Decodes data from RuuviTag with Data Format 3
    Protocol specification:
    https://github.com/ruuvi/ruuvi-sensor-protocols
    """

    def _get_temperature(self, data):
        """Return temperature in celsius"""

        # The temperature is in two fields, one for the integer part,
        # one for the fraction
        #
        # The integer part was decoded as a signed two's complement number,
        # but this isn't how it's really stored. The MSB is a sign, the lower
        # 7 bits are the unsigned temperature value.
        #
        # To convert from the decoded value we have to add 128 and then negate,
        # if the decoded value was negative
        frac = data[3] / 100
        if data[2] < 0:
            return -(data[2] + 128 + frac)
        else:
            return data[2] + frac

    def _get_humidity(self, data):
        """Return humidity %"""
        return data[1] * 0.5

    def _get_pressure(self, data):
        """Return air pressure hPa"""
        return (data[4] + 50000) / 100

    def _get_acceleration(self, data):
        """Return acceleration mG"""
        return data[5:8]

    def _get_battery(self, data):
        """Return battery mV"""
        return data[8]

    def decode_data(self, data):
        """
        Decode sensor data.

        Returns:
            dict: Sensor values
        """
        try:
            byte_data = struct.unpack(
                '>BBbBHhhhH', bytearray.fromhex(data[:28])
            )
            acc_x, acc_y, acc_z = self._get_acceleration(byte_data)
            return {
                'data_format': 3,
                'humidity': self._get_humidity(byte_data),
                'temperature': self._get_temperature(byte_data),
                'pressure': self._get_pressure(byte_data),
                'acceleration': math.sqrt(
                    acc_x * acc_x + acc_y * acc_y + acc_z * acc_z
                ),
                'acceleration_x': acc_x,
                'acceleration_y': acc_y,
                'acceleration_z': acc_z,
                'battery': self._get_battery(byte_data)
            }
        except Exception:
            log.exception('Value: %s not valid', data)
            return None


ruuvi_decoders/df5_decoder.py
from __future__ import division

import math
import logging
import struct

log = logging.getLogger(__name__)


class Df5Decoder(object):
    """
    Decodes data from RuuviTag with Data Format 5
    Protocol specification:
    https://github.com/ruuvi/ruuvi-sensor-protocols
    """

    def _get_temperature(self, data):
        """Return temperature in celsius"""
        if data[1] == -32768:
            return None

        return round(data[1] / 200, 2)

    def _get_humidity(self, data):
        """Return humidity %"""
        if data[2] == 65535:
            return None

        return round(data[2] / 400, 2)

    def _get_pressure(self, data):
        """Return air pressure hPa"""
        if data[3] == 0xFFFF:
            return None

        return round((data[3] + 50000) / 100, 2)

    def _get_acceleration(self, data):
        """Return acceleration mG"""
        if (data[4] == -32768 or data[5] == -32768 or data[6] == -32768):
            return (None, None, None)

        return data[4:7]

    def _get_powerinfo(self, data):
        """Return battery voltage and tx power"""
        battery_voltage = data[7] >> 5
        tx_power = data[7] & 0x001F

        return (battery_voltage, tx_power)

    def _get_battery(self, data):
        """Return battery mV"""
        battery_voltage = self._get_powerinfo(data)[0]
        if battery_voltage == 0b11111111111:
            return None

        return battery_voltage + 1600

    def _get_txpower(self, data):
        """Return transmit power"""
        tx_power = self._get_powerinfo(data)[1]
        if tx_power == 0b11111:
            return None

        return -40 + (tx_power * 2)

    def _get_movementcounter(self, data):
        return data[8]

    def _get_measurementsequencenumber(self, data):
        return data[9]

    def _get_mac(self, data):
        return ''.join('{:02x}'.format(x) for x in data[10:])

    def decode_data(self, data):
        """
        Decode sensor data.

        Returns:
            dict: Sensor values
        """
        try:
            byte_data = struct.unpack(
                '>BhHHhhhHBH6B', bytearray.fromhex(data[:48])
            )

            acc_x, acc_y, acc_z = self._get_acceleration(byte_data)
            return {
                'data_format': 5,
                'humidity': self._get_humidity(byte_data),
                'temperature': self._get_temperature(byte_data),
                'pressure': self._get_pressure(byte_data),
                'acceleration': math.sqrt(
                    acc_x * acc_x + acc_y * acc_y + acc_z * acc_z
                ),
                'acceleration_x': acc_x,
                'acceleration_y': acc_y,
                'acceleration_z': acc_z,
                'tx_power': self._get_txpower(byte_data),
                'battery': self._get_battery(byte_data),
                'movement_counter': self._get_movementcounter(byte_data),
                'measurement_sequence_number':
                    self._get_measurementsequencenumber(byte_data),
                'mac': self._get_mac(byte_data)
            }
        except Exception:
            log.exception('Value: %s not valid', data)
            return None


ruuvi_decoders/url_decoder.py
from __future__ import division

import base64
import logging

log = logging.getLogger(__name__)


class UrlDecoder(object):
    """
    Decodes data from RuuviTag url
    Protocol specification:
    https://github.com/ruuvi/ruuvi-sensor-protocols

    Decoder operations are ported from:
    https://github.com/ruuvi/sensor-protocol-for-eddystone-url/blob/master/index.html

    0:   uint8_t     format;          // (0x02 = realtime sensor readings)
    1:   uint8_t     humidity;        // one lsb is 0.5%
    2-3: uint16_t    temperature;     // Signed 8.8 fixed-point notation.
    4-5: uint16_t    pressure;        // (-50kPa)
    6-7: uint16_t    time;            // seconds (now from reset)

    The bytes for temperature, pressure and time are swaped during the encoding
    """

    def _get_temperature(self, decoded):
        """Return temperature in celsius"""
        temp = (decoded[2] & 127) + decoded[3] / 100
        sign = (decoded[2] >> 7) & 1
        if sign == 0:
            return round(temp, 2)
        return round(-1 * temp, 2)

    def _get_humidity(self, decoded):
        """Return humidity %"""
        return decoded[1] * 0.5

    def _get_pressure(self, decoded):
        """Return air pressure hPa"""
        pres = ((decoded[4] << 8) + decoded[5]) + 50000
        return pres / 100

    def decode_data(self, encoded):
        """
        Decode sensor data.

        Returns:
            dict: Sensor values
        """
        try:
            identifier = None
            data_format = 2
            if len(encoded) > 8:
                data_format = 4
                identifier = encoded[8:]
                encoded = encoded[:8]
            decoded = bytearray(base64.b64decode(encoded, '-_'))
            return {
                'data_format': data_format,
                'temperature': self._get_temperature(decoded),
                'humidity': self._get_humidity(decoded),
                'pressure': self._get_pressure(decoded),
                'identifier': identifier
            }
        except:
            log.exception('Encoded value: %s not valid', encoded)
            return None


ruuvi_decoders/decoders.py
"""Main module."""

from __future__ import division

import base64
import math
import logging

log = logging.getLogger(__name__)

# pylint: disable=no-self-use


def twos_complement(value, bits):
    if (value & (1 << (bits - 1))) != 0:
        value = value - (1 << bits)
    return value


def rshift(val, n):
    """
    Arithmetic right shift, preserves sign bit.
    https://stackoverflow.com/a/5833119 .
    """
    return (val % 0x100000000) >> n


class UrlDecoder(object):
    """
    Decodes data from RuuviTag url
    Protocol specification:
    https://github.com/ruuvi/ruuvi-sensor-protocols

    Decoder operations are ported from:
    https://github.com/ruuvi/sensor-protocol-for-eddystone-url/blob/master/index.html

    0:   uint8_t     format;          // (0x02 = realtime sensor readings)
    1:   uint8_t     humidity;        // one lsb is 0.5%
    2-3: uint16_t    temperature;     // Signed 8.8 fixed-point notation.
    4-5: uint16_t    pressure;        // (-50kPa)
    6-7: uint16_t    time;            // seconds (now from reset)

    The bytes for temperature, pressure and time are swaped during the encoding
    """

    def _get_temperature(self, decoded):
        """Return temperature in celsius"""
        temp = (decoded[2] & 127) + decoded[3] / 100
        sign = (decoded[2] >> 7) & 1
        if sign == 0:
            return round(temp, 2)
        return round(-1 * temp, 2)

    def _get_humidity(self, decoded):
        """Return humidity %"""
        return decoded[1] * 0.5

    def _get_pressure(self, decoded):
        """Return air pressure hPa"""
        pres = ((decoded[4] << 8) + decoded[5]) + 50000
        return pres / 100

    def decode_data(self, encoded):
        """
        Decode sensor data.

        Returns:
            dict: Sensor values
        """
        try:
            identifier = None
            data_format = 2
            if len(encoded) > 8:
                data_format = 4
                identifier = encoded[8:]
                encoded = encoded[:8]
            decoded = bytearray(base64.b64decode(encoded, '-_'))
            return {
                'data_format': data_format,
                'temperature': self._get_temperature(decoded),
                'humidity': self._get_humidity(decoded),
                'pressure': self._get_pressure(decoded),
                'identifier': identifier
            }
        except:
            log.exception('Encoded value: %s not valid', encoded)
            return None


class Df3Decoder(object):
    """
    Decodes data from RuuviTag with Data Format 3
    Protocol specification:
    https://github.com/ruuvi/ruuvi-sensor-protocols
    """

    def _get_temperature(self, data):
        """Return temperature in celsius"""
        temp = (data[2] & ~(1 << 7)) + (data[3] / 100)
        sign = (data[2] >> 7) & 1
        if sign == 0:
            return round(temp, 2)
        return round(-1 * temp, 2)

    def _get_humidity(self, data):
        """Return humidity %"""
        return data[1] * 0.5

    def _get_pressure(self, data):
        """Return air pressure hPa"""
        pres = (data[4] << 8) + data[5] + 50000
        return pres / 100

    def _get_acceleration(self, data):
        """Return acceleration mG"""
        acc_x = twos_complement((data[6] << 8) + data[7], 16)
        acc_y = twos_complement((data[8] << 8) + data[9], 16)
        acc_z = twos_complement((data[10] << 8) + data[11], 16)
        return (acc_x, acc_y, acc_z)

    def _get_battery(self, data):
        """Return battery mV"""
        return (data[12] << 8) + data[13]

    def decode_data(self, data):
        """
        Decode sensor data.

        Returns:
            dict: Sensor values
        """
        try:
            byte_data = bytearray.fromhex(data)
            acc_x, acc_y, acc_z = self._get_acceleration(byte_data)
            return {
                'data_format': 3,
                'humidity': self._get_humidity(byte_data),
                'temperature': self._get_temperature(byte_data),
                'pressure': self._get_pressure(byte_data),
                'acceleration': math.sqrt(
                    acc_x * acc_x + acc_y * acc_y + acc_z * acc_z),
                'acceleration_x': acc_x,
                'acceleration_y': acc_y,
                'acceleration_z': acc_z,
                'battery': self._get_battery(byte_data)
            }
        except Exception:
            log.exception('Value: %s not valid', data)
            return None


class Df5Decoder(object):
    """
    Decodes data from RuuviTag with Data Format 5
    Protocol specification:
    https://github.com/ruuvi/ruuvi-sensor-protocols
    """

    def _get_temperature(self, data):
        """Return temperature in celsius"""
        if data[1:2] == 0x7FFF:
            return None

        temperature = twos_complement((data[1] << 8) + data[2], 16) / 200
        return round(temperature, 2)

    def _get_humidity(self, data):
        """Return humidity %"""
        if data[3:4] == 0xFFFF:
            return None

        humidity = ((data[3] & 0xFF) << 8 | data[4] & 0xFF) / 400
        return round(humidity, 2)

    def _get_pressure(self, data):
        """Return air pressure hPa"""
        if data[5:6] == 0xFFFF:
            return None

        pressure = ((data[5] & 0xFF) << 8 | data[6] & 0xFF) + 50000
        return round((pressure / 100), 2)

    def _get_acceleration(self, data):
        """Return acceleration mG"""
        if (data[7:8] == 0x7FFF or
                data[9:10] == 0x7FFF or
                data[11:12] == 0x7FFF):
            return (None, None, None)

        acc_x = twos_complement((data[7] << 8) + data[8], 16)
        acc_y = twos_complement((data[9] << 8) + data[10], 16)
        acc_z = twos_complement((data[11] << 8) + data[12], 16)
        return (acc_x, acc_y, acc_z)

    def _get_powerinfo(self, data):
        """Return battery voltage and tx power"""
        power_info = (data[13] & 0xFF) << 8 | (data[14] & 0xFF)
        battery_voltage = rshift(power_info, 5) + 1600
        tx_power = (power_info & 0b11111) * 2 - 40

        if rshift(power_info, 5) == 0b11111111111:
            battery_voltage = None
        if (power_info & 0b11111) == 0b11111:
            tx_power = None

        return (round(battery_voltage, 3), tx_power)

    def _get_battery(self, data):
        """Return battery mV"""
        battery_voltage = self._get_powerinfo(data)[0]
        return battery_voltage

    def _get_txpower(self, data):
        """Return transmit power"""
        tx_power = self._get_powerinfo(data)[1]
        return tx_power

    def _get_movementcounter(self, data):
        return data[15] & 0xFF

    def _get_measurementsequencenumber(self, data):
        measurementSequenceNumber = (data[16] & 0xFF) << 8 | data[17] & 0xFF
        return measurementSequenceNumber

    def _get_mac(self, data):
        return ''.join('{:02x}'.format(x) for x in data[18:24])

    def decode_data(self, data):
        """
        Decode sensor data.

        Returns:
            dict: Sensor values
        """
        try:
            byte_data = bytearray.fromhex(data)
            acc_x, acc_y, acc_z = self._get_acceleration(byte_data)
            return {
                'data_format': 5,
                'humidity': self._get_humidity(byte_data),
                'temperature': self._get_temperature(byte_data),
                'pressure': self._get_pressure(byte_data),
                'acceleration': math.sqrt(acc_x * acc_x + acc_y * acc_y + acc_z * acc_z),
                'acceleration_x': acc_x,
                'acceleration_y': acc_y,
                'acceleration_z': acc_z,
                'tx_power': self._get_txpower(byte_data),
                'battery': self._get_battery(byte_data),
                'movement_counter': self._get_movementcounter(byte_data),
                'measurement_sequence_number': self._get_measurementsequencenumber(byte_data),
                'mac': self._get_mac(byte_data)
            }
        except Exception:
            log.exception('Value: %s not valid', data)
            return None
